xayn-ai/src/reranker/sync.rs

```diff
@@ -0,0 +1,157 @@
+#[cfg_attr(test, derive(Clone, PartialEq, Debug))]
+#[derive(Default, Serialize, Deserialize)]
+pub(crate) struct SyncData {
+    pub(crate) user_interests: UserInterests,
+}
+
+impl SyncData {
+    /// Deserializes a `SyncData` from `bytes`.
+    pub(crate) fn deserialize(bytes: &[u8]) -> Result<Self, Error> {
+        // version encoded in first byte
+        let version = bytes[0];
+        if version != CURRENT_SCHEMA_VERSION {
+            bail!(
+                "Unsupported serialized data. Found version {} expected {}.",
+                version,
+                CURRENT_SCHEMA_VERSION,
+            );
+        }
+
+        let data = bincode::deserialize(&bytes[1..])?;
+        Ok(data)
+    }
+
+    /// Serializes a `SyncData` to a byte representation.
+    pub(crate) fn serialize(&self) -> Result<Vec<u8>, Error> {
+        let size = bincode::serialized_size(self)? + 1;
+        let mut serialized = Vec::with_capacity(size as usize);
+        // version encoded in first byte
+        serialized.push(CURRENT_SCHEMA_VERSION);
+        bincode::serialize_into(&mut serialized, self)?;
+
+        Ok(serialized)
+    }
+
+    /// Synchronizes with another `SyncData`.
+    pub(crate) fn synchronize(&mut self, other: SyncData) {
+        let Self { user_interests } = other;
+        self.user_interests.append(user_interests);
+
+        reduce_cois(&mut self.user_interests.positive);
+        reduce_cois(&mut self.user_interests.negative);
+
+        self.user_interests.reassign_ids();
+    }
+}
+
+/// A pair of CoIs.
+#[derive(Clone)]
+struct CoiPair<C>(C, C);
+
+impl<C> CoiPair<C>
+where
+    C: CoiPoint,
+{
+    /// Merges the CoI pair, assigning it the smaller of the two ids.
+    fn merge(&self) -> C {
+        let CoiId(id1) = self.0.id();
+        let CoiId(id2) = self.1.id();
+        self.0.merge(&self.1, id1.min(id2))
+    }
+
+    /// True iff either CoI has the given id.
+    fn contains(&self, id: CoiId) -> bool {
+        self.0.id() == id || self.1.id() == id
+    }
+
+    /// True iff either CoI is one of the given pair.
+    fn contains_any(&self, other: &Self) -> bool {
+        self.contains(other.0.id()) || self.contains(other.1.id())
+    }
+}
+
+/// A `Coiple` is a pair of CoIs and the distance between them.
+#[derive(Clone)]
+struct Coiple<C> {
+    cois: CoiPair<C>,
+    dist: f32,
+}
+
+impl<C> Coiple<C> {
+    fn new(coi1: C, coi2: C, dist: f32) -> Self {
+        let cois = CoiPair(coi1, coi2);
+        Self { cois, dist }
+    }
+}
+
+/// Computes the l2 distance between two CoI points.
+fn dist<C>(coi1: &C, coi2: &C) -> f32
+where
+    C: CoiPoint,
+{
+    l2_distance(coi1.point(), coi2.point())
+}
+
+/// Reduces the given collection of CoIs by successively merging the pair in closest
+/// proximity to each other.
+///
+/// <https://xainag.atlassian.net/wiki/spaces/XAY/pages/2029944833/CoI+synchronisation>
+/// outlines the core of the algorithm.
+fn reduce_cois<C>(cois: &mut Vec<C>)
```

Comment:
yes, will follow up this todo in the next.

Reply:
Can we add this precondition to the description of the function?

Reply:
should still work in the presence of dupes (albeit with extra wasted work) so prob not necessary to make it a precondition. although in practice, it will be free from dupes (once that is added).